packages/core/src/components/NumberField/hooks/useNumberFieldState.ts

```diff
@@ -0,0 +1,124 @@
+import { useCallback, useEffect, useState } from "react";
+import { NumberFieldProps } from "../NumberField.types";
+import { calculateSteppedValue } from "../utils/calc-value";
+
+export type UseNumberFieldStateProps = Pick<
+  NumberFieldProps,
+  "value" | "onChange" | "min" | "max" | "step" | "disabled" | "readOnly" | "allowOutOfBounds" | "onValidityChange"
+>;
+
+const useNumberFieldState = ({
+  value: controlledValue,
+  onChange,
+  min,
+  max,
+  step = 1,
+  disabled,
+  readOnly,
+  allowOutOfBounds,
+  onValidityChange
+}: UseNumberFieldStateProps) => {
+  const [inputValue, setInputValue] = useState(controlledValue === null ? "" : String(controlledValue));
+
+  useEffect(() => {
+    const controlledValueStr = controlledValue === null ? "" : String(controlledValue);
+    if (controlledValue !== parseFloat(inputValue)) {
```

Comment:
This is tricky, there are many edge cases with number conversion in JS, I'd use a library for that